### PR TITLE
Added IEcsSystem management in the Configurator class

### LINKED ISSUE #502: `NovelRT::Ecs::Configurator` appears to be missing `WithSystems` overload for `IEcsSystem` implementors.
The `Configurator` helper class appears to only support the classic `std::function<void(TimeStamp, Catalogue)>` signature at the moment, but the `SystemScheduler` has support for also registering `IEcsSystem` objects. We should extend the `Configurator` to support this new overload.

### PR DESCRIPTION
**Pull request by bjauny**
**Please check if the PR fulfills these requirements**
- [x] The commit message follows our [guidelines](https://github.com/novelrt/NovelRT/blob/misc/templates/Contributing.md#pull-requests)
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Documentation has been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
This adds support for the `IEcsSystem` use in the `Configurator` class.

**Is there an open issue that this resolves? If so, please [link it here](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).**
Fixes #502 .

**What is the current behavior?** (You can also link to an open issue here)
`Configurator` only manages `std::function<void(TimeStamp, Catalogue)>` systems.

**What is the new behavior (if this is a feature change)?**
`Configurator` manages `IEcsSystem` systems.

**Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)
No.

**Other information**:

### CODE REVIEW

**Comment by bjauny** [comment]
This is a first and naive implementation.

Now that I reread the original issue, I think I should just rename the `WithIEcsSystems` to an overload of `WithSystems`. If so, I will add another test to manage the initalization with the overload instead of managing both in the current test.

**Comment by RubyNova** [comment]
I will wait for your changes to do a review then to avoid duplication. :D

**Comment by bjauny** [comment]
> I will wait for your changes to do a review then to avoid duplication. :D

All done!

File: include/NovelRT/Ecs/Configurator.h
Comment (RubyNova) [inline include/NovelRT/Ecs/Configurator.h:1]: Please remove this, it has the potential to break due to how our headers are structured. If it's causing problems to remove I can probably walk you through how to fix it correctly.

Looks like there's an additional include here that should not exist and a missing namespace error preprocessor error we have in our headers, maybe we should fix that while we're here.

File: include/NovelRT/Ecs/Configurator.h
Comment (RubyNova) [inline include/NovelRT/Ecs/Configurator.h:93]: The parameter name is wrong here.

File: include/NovelRT/Ecs/Configurator.h
Comment (RubyNova) [inline include/NovelRT/Ecs/Configurator.h:91]:
```suggestion
         * @brief Configures systems for registration for this ECS instance (IEcsSystem variant).
```
We do this to allow better doxygen linkage.

File: tests/NovelRT.Tests/Ecs/ConfiguratorTest.cpp
Comment (RubyNova) [inline tests/NovelRT.Tests/Ecs/ConfiguratorTest.cpp:58]: This could just be `auto` given the fact you're calling make shared. No reason to have type information twice unless you're explicitly changing said type I don't think.

File: include/NovelRT/Ecs/Configurator.h
Comment (RubyNova) [inline include/NovelRT/Ecs/Configurator.h:25]: I do wonder of there's a better name for this collection but it does what it says on the tin and I can't think of anything better right now so best just leave it alone. Mostly just writing this for transparency in case other maintainers see this and have an idea.

**Review by RubyNova — CHANGES_REQUESTED** [review]
Basically almost there, just some typos and admin stuff I think. :)

File: include/NovelRT/Ecs/Configurator.h
Comment (bjauny) [reply include/NovelRT/Ecs/Configurator.h:1]: Is it just a matter of including `Ecs.h` explicitely and remove the `#error`?

Comment (RubyNova) [reply include/NovelRT/Ecs/Configurator.h:1]: Remove the include, add the error define unless it's already in Configurator.h and I missed it?

File: tests/NovelRT.Tests/Ecs/ConfiguratorTest.cpp
Comment (bjauny) [reply tests/NovelRT.Tests/Ecs/ConfiguratorTest.cpp]: Done.

File: include/NovelRT/Ecs/Configurator.h
Comment (bjauny) [reply include/NovelRT/Ecs/Configurator.h]: Done.

File: include/NovelRT/Ecs/Configurator.h
Comment (bjauny) [reply include/NovelRT/Ecs/Configurator.h]: Done.

File: include/NovelRT/Ecs/Configurator.h
Comment (bjauny) [reply include/NovelRT/Ecs/Configurator.h]: Done. There is indeed a `#error` if the ECS define is not set, is that what you're expecting? The same can be found in 28 other files.

Comment (RubyNova) [reply include/NovelRT/Ecs/Configurator.h]: Yeah I see it now.

There's still an include we should remove here but I'll flag it in the second pass. :)

Comment (RubyNova) [reply include/NovelRT/Ecs/Configurator.h]: Ok nvm GitHub mobile won't let me - can you remove the `#include utility` at the top? The very first thing should be the file comment and there should be no additional includes in here.

Basically just fixing something that should not have been a problem in the first place.

**Review by RubyNova — CHANGES_REQUESTED** [review]
The only change I ask for is the one I mentioned in a previous comment.

I'll approve once that's in. :)

**Comment by bjauny** [comment]
And we're in!

**Review by RubyNova — APPROVED** [review]
Pizza party time! I'm happy with it, LGTM!

**Pull request merged**
